### DISCUSSION TITLE
Floor plan layer sun/cloud effect with opacity

### DISCUSSION THREAD

**Discussion by Trzinka** [ExperienceLovelace/ha-floorplan#410] [Q&A] [answered]
Hello.

# I had a problem when I made the image change depending on the time of day.

### when the external situation was completely different (cloudy).

`Example image at noon:`
![popoldan](https://github.com/ExperienceLovelace/ha-floorplan/assets/40424965/a1c30874-511c-47cd-855c-464394d3e2cc)

`A more realistic display:`
![jutro](https://github.com/ExperienceLovelace/ha-floorplan/assets/40424965/f9e5c5ec-4c5a-43e8-b450-1e21bfa2e15e)

**First I tried the code for Cloud coverage:**
```yaml
  - platform: template
    sensors:
      cloud_coverage:
        friendly_name: "Cloud coverage"
        value_template: >-
          {% set elevation = state_attr('sun.sun','elevation') | float %}
          {% set cloud_coverage = states('sensor.astroweather_cloud_cover') | float %}
          {% set cloud_factor = (1 - (0.75 * ( cloud_coverage / 100) ** 3 )) %}
          {{ cloud_factor | round(2) }}
        unit_of_measurement: '%'
        device_class: 'illuminance'
```
**Since I don't have a local external Illuminance sensor (lx) yet, I used sensor.astroweather_cloud_cover, which does not show the real situation where I live (distance from the capital Ljubljana, which is the only one I can choose from the list).**
![Cloud coverage](https://github.com/ExperienceLovelace/ha-floorplan/assets/40424965/72ccc009-a7e9-4431-b75b-7136dd5763a9)

`So I made another sensor for Sun brightness:`
```yaml 
sun_brightness:
        friendly_name: "Sun brightness"
        value_template: >-
          {% set min_bright_percent = 50 %}
          {% set max_bright_percent = 100 %}
          {% set mins_span = 30 %}
          {% set sr = today_at((state_attr('sun.sun', 'next_rising') | as_datetime | as_local).strftime('%H:%M')) %}
          {% set pre_sunrise =  sr - timedelta(minutes=mins_span ) %}
          {% set post_sunrise = sr + timedelta(minutes=mins_span ) %}
          {% set ss = today_at((state_attr('sun.sun', 'next_setting') | as_datetime | as_local).strftime('%H:%M')) %}
          {% set pre_sunset = ss - timedelta(minutes=mins_span ) %}
          {% set post_sunset = ss + timedelta(minutes=mins_span ) %}
          {% set t = now() %}
          {{ { pre_sunrise <= t < post_sunrise: max_bright_percent - min_bright_percent + ((t - pre_sunrise).seconds * (max_bright_percent - min_bright_percent )/3600)|int(0),
              post_sunrise <= t < pre_sunset: max_bright_percent,
              pre_sunset <= t < post_sunset: max_bright_percent - ((t - pre_sunset).seconds * (max_bright_percent - min_bright_percent )/3600)|int(0)
            }.get(true, min_bright_percent) }}  
        unit_of_measurement: '%'
        device_class: 'illuminance'
```
**To get the position of the sun and Sun brightness**
![Sun brightness](https://github.com/ExperienceLovelace/ha-floorplan/assets/40424965/263a7b53-d0c1-459b-82ac-9570217c4954)

**After that I also like to looked at the UV sensor:**
![UV-index](https://github.com/ExperienceLovelace/ha-floorplan/assets/40424965/178abdc2-a6bc-47ba-8d32-63f791db4255)

**I have an internal sensor with a built-in meter (lx):**
![PIR-Lx](https://github.com/ExperienceLovelace/ha-floorplan/assets/40424965/92ed7439-acda-4714-9a9f-d74978120c82)

**At the same time (evening/night) I also watched:**
![Astro isti čas kot Windy](https://github.com/ExperienceLovelace/ha-floorplan/assets/40424965/f7102ef3-f3d7-4642-a56f-2c0e90db5f2d)

![Windy](https://github.com/ExperienceLovelace/ha-floorplan/assets/40424965/2e1ded4f-bc63-4812-adaa-a52ae0c5425f)

**Now the outside looks somewhat comparable to this:**
![noc](https://github.com/ExperienceLovelace/ha-floorplan/assets/40424965/f11720cf-d3c0-4902-a07f-5c8ab7375beb)

## I have a desire and would like to ask if someone is willing to write me a code that would calculate the transparency from the given parameters.

`Which I would then use in the floor plan for the layer. `

```yaml 
- element: background
  entity: sensor.sunlight_opacity
  tap_action: false
  hover_action: false
  state_action:
    action: call-service
    service: floorplan.style_set
    service_data: |
      >
        var opacity = '';
        switch (entity.state) { 
         case '0.0': opacity ='0.1'; break; 
         case '0.1': opacity ='0.1'; break; 
         case '0.2': opacity ='0.2'; break; 
         case '0.3': opacity ='0.3'; break; 
         case '0.4': opacity ='0.4'; break; 
         case '0.5': opacity ='0.5'; break; 
         case '0.6': opacity ='0.6'; break; 
         case '0.7': opacity ='0.7'; break; 
         case '0.8': opacity ='0.8'; break; 
         case '0.9': opacity ='0.9'; break; 
         case '1.00': opacity ='1'; break; 
          }
      return `opacity: ${opacity};`;
```

Thank you.

**Comment by OzGav** [comment]
Which given parameters? If you mean all of the above then you would have to provide an algorithm that explains when one should be used over another or if you want the sensors combined then how would you want them combined?

**Reply by Trzinka** [reply]
# Let me think!

`The situation at 12:15 was as follows:`
![Sensors at 12-15](https://github.com/ExperienceLovelace/ha-floorplan/assets/40424965/8e11802c-698a-4ca9-8e51-f28bf50fb2d9)

`The picture is quite realistic:`
![Windy at 12-15](https://github.com/ExperienceLovelace/ha-floorplan/assets/40424965/103fbf4d-154d-4e5a-9daf-7dc6e7116e9e)

### By the way, all the code that is written is not mine, maybe I changed something (trial/error) and to be honest, I don't know this part very well!

**Reply by Trzinka** [reply]
## Let's look at it this way:

### I wouldn't use Sun brightness for now because it's 0 or 100 most of the time
```
Cloud coverage = from 0 to 1
UV-index = from 0 to 11
PIR-Lx = from 0 to 100 (in all likelihood)!
```
## I think it would be best like this:
```
Cloud coverage rounded to 1 decimal place! 0.x5 up 0.x4 down rounded
PIR-Lx divided with 100 and rounded to 1 decimal place! 0.x5 up 0.x4 down rounded
UV-index divided with 10 unless 11 = 1
```

Now I would calculate the average from these calculations!

# Based on image 0.5, 0.3 and 0.7, the average would be 0.5

**Reply by Trzinka** [reply]
## O! Sorry, I forgot:
```
Cloud coverage = sensor.cloud_coverage
PIR-Lx = sensor.pir_je_illuminance_lux
UV-index = sensor.weatherapi_dom_uv_index
```

**Reply by OzGav** [reply]
I can't test this but see if this works. Look in the browser console for errors if it doesn't. I also put a log line in there which I hope works which will show you the value getting returned.
```
- element: background
  tap_action: false
  hover_action: false
  state_action:
    service: floorplan.style_set
    service_data: |
      >
        var opacity = parseFloat((hass.states["sensor.cloud_coverage"].state +
                      Math.floor(hass.states["sensor.pir_je_illuminance_lux"].state / 100) +
                      Math.min(1,Math.floor(hass.states["sensor.weatherapi_dom_uv_index"].state / 10)) +
                      ) / 3 ).toFixed(1);
      console.log('opacity', ${opacity});              
      return `opacity: ${opacity};`;
```

**Reply by Trzinka** [reply]
### Hi! 
## Thanks for an idea.

## I'm somehow solving the matter and I encountered a banal problem when the only time in the calculation/average the result comes to 2 decimal places. ;=((((
![Final](https://github.com/ExperienceLovelace/ha-floorplan/assets/40424965/a146f71b-8b3b-46e1-9c5e-93364cfaa51d)


`Templets that I made:`

```
  - platform: template
    sensors:
      cloud_coverage:
        friendly_name: "Cloud coverage"
        value_template: >-
          {% set elevation = state_attr('sun.sun','elevation') | float %}
          {% set cloud_coverage = states('sensor.astroweather_cloud_cover') | float %}
          {% set cloud_factor = (1 - (0.75 * ( cloud_coverage / 100) ** 3 )) %}
          {{ cloud_factor | round(1) }}
###############
      pir_illuminance:
        friendly_name: "PIR osvetljenost"
        value_template: >-
          {% set pir_illuminance = states('sensor.pir_je_illuminance_lux') | float / 100 %}
          {{ pir_illuminance | round(1) }}
##
      uv_index_pretvorba:
        friendly_name: "UV index pretvorba"
        value_template: >-
          {% set uv_index = states('sensor.weatherapi_dom_uv_index') | float / 10 %}
          {% if uv_index == 11 %}
            {% set uv_opacity = 1 %}
          {% else %}
            {% set uv_opacity = (uv_index * 1) | round(1) %}
          {% endif %}
          {{ uv_opacity | round(1) }}

  - platform: average
    name: 'Povprecje zatemnitve doma'
    entities:
      - sensor.cloud_coverage
      - sensor.pir_illuminance
      - sensor.uv_index_pretvorba
```
`in floor plan card:`

```
 - element: bg
              entity: sensor.povprecje_zatemnitve_doma
              tap_action: false
              hover_action: false
              state_action:
                action: call-service
                service: floorplan.style_set
                service_data: |
                  >
                    var opacity = '';
                    switch (entity.state) { 
                     case '0.0': opacity ='0.1'; break; 
                     case '0.1': opacity ='0.1'; break; 
                     case '0.2': opacity ='0.2'; break; 
                     case '0.3': opacity ='0.3'; break; 
                     case '0.4': opacity ='0.4'; break; 
                     case '0.5': opacity ='0.5'; break; 
                     case '0.6': opacity ='0.6'; break; 
                     case '0.7': opacity ='0.7'; break; 
                     case '0.8': opacity ='0.8'; break; 
                     case '0.9': opacity ='0.9'; break; 
                     case '1.00': opacity ='1'; break; 
                      }
                    return `opacity: ${opacity};`;
```

## Now I just have to figure out how to round to 1 decimal for the average sensor!!!!

# Any idea?

**Reply by OzGav** [reply]
Try changing `entity.state` to `parseFloat(entity.state).toFixed(1)`

**Reply by OzGav** [reply, marked as answer]
Actually you don’t even need to do that as you could just set the opacity to the entity state as your switch/case is doing nothing

so 
```
return `opacity: ${entity.state};`;

```

**Reply by Trzinka** [reply]
## Thanks!

`The end result:`
![FP-Opacity](https://github.com/ExperienceLovelace/ha-floorplan/assets/40424965/6876df08-26d2-42f1-ac50-f1a8e6f85c8c)

`I hope it's not a coincidence that right now the average is at 1 decimal place.`
![Floor plan-Opacity](https://github.com/ExperienceLovelace/ha-floorplan/assets/40424965/0c0284ad-5c2f-432c-ba93-b925a2099e46)

![Povprečna zatemnitev doma](https://github.com/ExperienceLovelace/ha-floorplan/assets/40424965/6c679a9c-afa0-4d92-9749-ea168c915b92)


`The final code should be for the floor plan card:`

```
- element: bg
              entity: sensor.povprecje_zatemnitve_doma
              tap_action: false
              hover_action: false
              state_action:
                action: call-service
                service: floorplan.style_set
                service_data: |
                  >
                    var opacity = '';
                    switch (entity.state) { 
                     case '0.0': opacity ='0.1'; break; 
                     case '0.1': opacity ='0.1'; break; 
                     case '0.2': opacity ='0.2'; break; 
                     case '0.3': opacity ='0.3'; break; 
                     case '0.4': opacity ='0.4'; break; 
                     case '0.5': opacity ='0.5'; break; 
                     case '0.6': opacity ='0.6'; break; 
                     case '0.7': opacity ='0.7'; break; 
                     case '0.8': opacity ='0.8'; break; 
                     case '0.9': opacity ='0.9'; break; 
                     case '1.00': opacity ='1'; break; 
                      }
                    return `opacity: ${entity.state};`;
```

**Reply by OzGav** [reply]
Remove the entire switch statement as it is doing nothing.

**Reply by Trzinka** [reply]
`Did you mean so:`
```
- element: bg
  entity: sensor.povprecje_zatemnitve_doma
  tap_action: false
  hover_action: false
  state_action:
    action: call-service
    service: floorplan.style_set
    service_data: |
      >
       var opacity = '';
         return `opacity: ${entity.state};`;
```

**Reply by OzGav** [reply]
Yes but you also remove the var opacity line and the | and > and put the return on the service_data line

**Reply by Trzinka** [reply]
`Did you mean:`
```
 - element: bg
              entity: sensor.povprecje_zatemnitve_doma
              tap_action: false
              hover_action: false
              state_action:
                action: call-service
                service: floorplan.style_set
                service_data: |
                  > return `opacity: ${entity.state};`
```

**Reply by Trzinka** [reply]
## By the way, it also works if it is sensor.povprecje_zatemnitve_doma with two decimal places!!

**Reply by OzGav** [reply]
Correct

**Reply by OzGav** [reply]
You can make it even shorter

```
              element: bg
              entity: sensor.povprecje_zatemnitve_doma
              tap_action: false
              hover_action: false
              state_action:
                service: floorplan.style_set
                service_data: return `opacity: ${entity.state};`
```

**Reply by Trzinka** [reply]
`I had to fix it so that it is now like this:`
```
- element: background
  entity: sensor.povprecje_zatemnitve_doma
  tap_action: false
  hover_action: false
  state_action:
    action: call-service
    service: floorplan.style_set
    service_data: |
      > return `opacity: ${entity.state};`
```

`Because of CSS settings:`
```
#background {
  display: block !important;
  z-index: 1;	
 }
```

### In order to enable room lighting without the influence of the sensor.povprecje_zatemnitve_doma.
`The result is as follows:`
![FP-Switch ON](https://github.com/ExperienceLovelace/ha-floorplan/assets/40424965/72f457cc-2b21-40bc-924b-713661d63c10)

`In CSS settings I use it to illuminate rooms and borders:`
```
#rooms image{
	display: none;
	mix-blend-mode: lighten;
	animation: 0.75s fade-in linear forwards;
  }

  @keyframes fade-in {
	from {
		opacity: 0;
	}
	to {
		opacity: 1;
	}
  }

#rooms image.switch-on {
	display: block !important;
	opacity: 1;
	z-index: 500 !important;	
  }

#areas {
	fill: none;
	stroke: none;
  }

#areas .switch-on {
	stroke: red !important;
	stroke-width: 30px !important;
	stroke-opacity: 1;
	z-index: 600;
  }
 }
```

**Reply by Trzinka** [reply]
`I tried as you wrote, but I can't make it any shorter than that:`
```yaml
- element: background
  entity: sensor.povprecje_zatemnitve_doma
  tap_action: false
  hover_action: false
  state_action:
    service: floorplan.style_set
    service_data: |
      > return `opacity: ${entity.state}`
```

# Thank you very much. You are a magic man!!!
